### ISSUE TITLE
config file

### ISSUE THREAD

**Issue by mock1328** [apognu/tuigreet#156]
IMO, would be great if we had a `/etc/greetd/tuigreet.toml` config file alongside running with args.
Options line gets too long and ugly very quickly, managing config file could be easier

**Comment by apognu** [comment]
Thanks for the issue. I was actually thinking about doing something like it, since this also bothers me.

I'll categorize this as a feature request and I'll probably work on it for 0.11.

**Comment by apognu** [comment]
This is being worked on and, on the relevant branch, it is at a stage where it can be _tested_ and polished. I'll still ask that, if someone wanted to give it a go, they be careful, it is still very much a work in progress.

I plan on having this ready for 0.11 (so _not_ the next version).

**Comment by Hugo-Heagren** [comment]
I have used tuigreet from NixOS before, and have just set it up again (thus finding my way here, finding to my great delight this had already been handled :smiley:). It would be really good if there was a structured way of setting configs from the `configuration.nix` -- more structured than just writing a string. Has that been thought of? Should I open a new issue?

I'm imaging that something like this would write the config file in the relevant way (perhaps overwriting everything previously, perhaps only setting the keys which are explicit set herein):

``` nix
services.greetd = {
  enable = true;
  settings = {
    default_session = {
      command = "${lib.makeBinPath [pkgs.greetd.tuigreet] }/tuigreet";
      user = "greeter";
    }; 
    vt = 1;
  };
 # Not sure where the best place for this would be. Perhaps not inside greetd? IDK.
  tuigreet.config = {
    time = true;
    asterisks = true;
    command = "bash";
    # Note structured theme, I think this would be handy for applying
    # system-wide colour schemes
    theme = {
      border = magenta;
      text = cyan;
      prompt = green;
      time = red;
      action = blue;
      button = yellow;
      container = black;
      input = red;
    };
  };
};

```

**Comment by apognu** [comment]
@Hugo-Heagren Yes, this is ongoing (albeit taking a bit more time than I anticipated) and will use TOML indeed.

As far as custom Nix derivations for `tuigreet` config, I am not really sure what I can do here since I do not maintain the packaging for various distributions.

**Comment by Brisingr05** [comment]
@Hugo-Heagren If/when TOML config file support is added in tuigreet, changes would need to be made in the Nixpkgs repo, rather than this one. Separate `programs.tuigreet.*` options will need to be created, like for [ReGreet](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=programs.regreet), another greetd greeter. TOML is a Nix-representable format, so `pkgs.formats.toml` can be used ([relevant manual section](https://nixos.org/manual/nixos/unstable/#sec-settings-nix-representable)). `pkgs.formats.*` is used for the `settings` option in various NixOS and Home Manager modules.

For a slightly cleaner config (in the current state), I use the `concatStringsSep` function:

```nix
{ lib, pkgs, ... }:

let
  inherit (lib) getExe concatStringsSep;
in

{
  services.greetd = {
    enable = true;
    settings = {
      default_session = {
        command = concatStringsSep " " [
          (getExe pkgs.greetd.tuigreet)
          "--time"
          "--asterisks"
          "--cmd 'bash'"
          "--theme ${
            concatStringsSep ";" [
              "border=magenta"
              "text=cyan"
              "prompt=green"
              "time=red"
              "action=blue"
              "button=yellow"
              "container=black"
              "input=red"
            ]
          }"
        ];
        user = "greeter";
      };
      vt = 1;
    };
  };
}
```

**Comment by antoineco** [comment]
This feature was added in 68b460d. Thank you!
